### PR TITLE
transform request-options on redirect

### PR DESCRIPTION
**Pull request by claasahl**
This PR is related to #114 

It allows one to optionally change the request-options on every redirect.

### CODE REVIEW

File: test/test-with-server.js [hunk 1]
```diff
@@ -1348,6 +1348,108 @@ describe("follow-redirects", function () {
         });
     });
   });
+
+  describe("transform request options", function () {
+    it("don't call transform function", function () {
```
Comment (RubenVerborgh) [inline test/test-with-server.js]: Test description unclear

File: test/test-with-server.js [hunk 2]
```diff
@@ -1348,6 +1348,108 @@ describe("follow-redirects", function () {
...
+          };
+          http.get(options, concatJson(resolve, reject)).on("error", reject);
+        }))
+        .then(function (res) {
+          assert.strictEqual(callsToTransform, 0);
+          assert.deepEqual(res.parsedJson, { a: "b" });
+          assert.deepEqual(res.responseUrl, "http://localhost:3600/a");
+        });
+    });
+    it("recover, if transform function is null", function () {
```
Comment (RubenVerborgh) [inline test/test-with-server.js]: "ignore beforeRedirect if not a function"

and only one test is needed, just set `beforeRedirect` to `3` or so.

File: test/test-with-server.js [hunk 3]
```diff
@@ -1348,6 +1348,108 @@ describe("follow-redirects", function () {
...
+            transformOptions: null,
+          };
+          http.get(options, concatJson(resolve, reject)).on("error", reject);
+        }))
+        .then(function (res) {
+          assert.deepEqual(res.parsedJson, { a: "b" });
+          assert.deepEqual(res.responseUrl, "http://localhost:3600/c");
+        });
+    });
+    it("recover, if transform function is undefined", function () {
```
Comment (RubenVerborgh) [inline test/test-with-server.js]: this test can be removed

File: test/test-with-server.js [hunk 4]
```diff
@@ -1348,6 +1348,108 @@ describe("follow-redirects", function () {
...
+            method: "GET",
+            transformOptions: undefined,
+          };
+          http.get(options, concatJson(resolve, reject)).on("error", reject);
+        }))
+        .then(function (res) {
+          assert.deepEqual(res.parsedJson, { a: "b" });
+          assert.deepEqual(res.responseUrl, "http://localhost:3600/c");
+        });
+    });
```
Comment (RubenVerborgh) [inline test/test-with-server.js]: empty line between tests

File: test/test-with-server.js [hunk 5]
```diff
@@ -1348,6 +1348,108 @@ describe("follow-redirects", function () {
...
+    it("transform on both redirects", function () {
```
Comment (RubenVerborgh) [inline test/test-with-server.js]: Clearer test description

File: test/test-with-server.js [hunk 6]
```diff
@@ -1348,6 +1348,108 @@ describe("follow-redirects", function () {
...
+          http.get(options, concatJson(resolve, reject)).on("error", reject);
+        }))
+        .then(function (res) {
+          assert.strictEqual(callsToTransform, 2);
+          assert.strictEqual(res.parsedJson["header-a"], "value A");
+          assert.strictEqual(res.parsedJson["header-b"], "value B");
+          assert.deepEqual(res.responseUrl, "http://localhost:3600/c");
+        });
+    });
+  });
```
Comment (RubenVerborgh) [inline test/test-with-server.js]: Plus a test for when the callback throws an error; should abort the entire request chain with that error.

File: README.md [hunk 7]
```diff
@@ -45,13 +45,21 @@ Global options are set directly on the `follow-redirects` module:
...
+  }
+};
 ```
 
 The following global options are supported:
 
 - `maxRedirects` (default: `21`) – sets the maximum number of allowed redirects; if exceeded, an error will be emitted.
 
 - `maxBodyLength` (default: 10MB) – sets the maximum size of the request body; if exceeded, an error will be emitted.
+-
```
Comment (RubenVerborgh) [inline README.md]: Remove

File: README.md [hunk 8]
```diff
@@ -45,13 +45,21 @@ Global options are set directly on the `follow-redirects` module:
 var followRedirects = require('follow-redirects');
 followRedirects.maxRedirects = 10;
 followRedirects.maxBodyLength = 20 * 1024 * 1024; // 20 MB
+followRedirects.transformOptions = function (options) {
```
[the remark this answers is not in the document]
Comment (claasahl) [reply README.md]: done

File: README.md [hunk 9]
```diff
@@ -45,13 +45,21 @@ Global options are set directly on the `follow-redirects` module:
...
+  // this function is called on every redirect
```
[the remark this answers is not in the document]
Comment (claasahl) [reply README.md]: done

File: README.md [hunk 10]
```diff
@@ -45,13 +45,21 @@ Global options are set directly on the `follow-redirects` module:
...
+  if(options.hostname === "www.google.com") {
```
[the remark this answers is not in the document]
Comment (claasahl) [reply README.md]: done

File: README.md [hunk 11]
```diff
@@ -45,13 +45,21 @@ Global options are set directly on the `follow-redirects` module:
...
+    options.headers["special-header"] = "special value";
```
[the remark this answers is not in the document]
Comment (claasahl) [reply README.md]: hope this is a more meaningful example 😄 

alternatively, I can provide examples which cover things like: using alternative HTTP methods for redirects, setting timeouts or changing the agent

File: README.md [hunk 12]
```diff
@@ -45,13 +45,21 @@ Global options are set directly on the `follow-redirects` module:
...
+};
 ```
 
 The following global options are supported:
 
 - `maxRedirects` (default: `21`) – sets the maximum number of allowed redirects; if exceeded, an error will be emitted.
 
 - `maxBodyLength` (default: 10MB) – sets the maximum size of the request body; if exceeded, an error will be emitted.
+- 
+- `transformOptions` (default: no operation) – optionally change request `options`; this function is called on every redirect.
```
[the remark this answers is not in the document]
Comment (claasahl) [reply README.md]: done

File: index.js [hunk 13]
```diff
@@ -349,6 +357,7 @@ function wrap(protocols) {
   var exports = {
     maxRedirects: 21,
     maxBodyLength: 10 * 1024 * 1024,
+    transformOptions: noop,
```
[the remark this answers is not in the document]
Comment (claasahl) [reply index.js]: no, not needed. I removed this.

File: test/test-with-server.js (hunk 2 above)
Comment (claasahl) [reply test/test-with-server.js]: done

File: test/test-with-server.js (hunk 3 above)
Comment (claasahl) [reply test/test-with-server.js]: done

File: test/test-with-server.js (hunk 4 above)
Comment (claasahl) [reply test/test-with-server.js]: done

File: test/test-with-server.js (hunk 5 above)
Comment (claasahl) [reply test/test-with-server.js]: hope this is better

File: test/test-with-server.js (hunk 6 above)
Comment (claasahl) [reply test/test-with-server.js]: done

File: README.md (hunk 7 above)
Comment (claasahl) [reply README.md]: done

File: index.js [hunk 14]
```diff
@@ -18,6 +18,31 @@ var eventHandlers = Object.create(null);
   };
 });
 
+function cleanupOptions(options) {
```
[the remark this answers is not in the document]
Comment (claasahl) [reply index.js]: done

File: index.js [hunk 15]
```diff
@@ -326,6 +330,10 @@ RedirectableRequest.prototype._processResponse = function (response) {
     var redirectUrl = url.resolve(this._currentUrl, location);
     debug("redirecting to", redirectUrl);
     Object.assign(this._options, url.parse(redirectUrl));
+    if (typeof this._options.transformOptions === "function") {
+      this._options.transformOptions.apply(null, [this._options]);
```
[the remark this answers is not in the document]
Comment (claasahl) [reply index.js]: very good point. thanks.

File: index.js [hunk 16]
```diff
@@ -388,6 +397,7 @@ function wrap(protocols) {
       options = Object.assign({
         maxRedirects: exports.maxRedirects,
         maxBodyLength: exports.maxBodyLength,
+        transformOptions: exports.transformOptions,
```
[the remark this answers is not in the document]
Comment (claasahl) [reply index.js]: done

File: test/test-with-server.js (hunk 1 above)
Comment (claasahl) [reply test/test-with-server.js]: updated description

File: index.js [hunk 17]
```diff
@@ -53,18 +44,6 @@ function RedirectableRequest(options, responseCallback) {
...
-    var searchPos = options.path.indexOf("?");
-    if (searchPos < 0) {
-      options.pathname = options.path;
-    }
-    else {
-      options.pathname = options.path.substring(0, searchPos);
-      options.search = options.path.substring(searchPos);
-    }
-  }
-
```
Comment (RubenVerborgh) [inline index.js]: Moving this logic before `_processResponse` is incorrect, because  `_processResponse` can still change `path`. Will investigate.

**Comment by RubenVerborgh** [comment]
Thanks @claasahl!

**Pull request merged**